src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -24,7 +25,13 @@ internal CachingContext CacheContext
             get
             {
                 Debug.Assert(IsReadOnly);
-                return _cachingContext ??= TrackedCachingContexts.GetOrCreate(this);
+                return _cachingContext ?? GetOrCreate();
+
+                CachingContext GetOrCreate()
+                {
+                    CachingContext ctx = TrackedCachingContexts.GetOrCreate(this);
+                    return Interlocked.CompareExchange(ref _cachingContext, ctx, null) ?? ctx;
```

Comment:
Is there any cleanup that should be done on ctx if this loses the race?

Reply:
If it was added to the shared cache then it will still get collected since it's stored there as a weak reference. And because access to the shared cache is synchronized the competing thread will have returned the same instance.